README.md
# uarizona-salary-app

Streamlit app to illustrate University of Arizona salary data.

Deployed as Heroku at: https://sapp4ua.herokuapp.com

**TL;DR:**<br>
_This is a website providing public salary data for the University of
Arizona. It is a "Choose Your Own Data Science" tool, so just
explore with different "data views" on the sidebar!_

**More information:**<br>
This site was developed to allow easy extraction, analysis, and interpretation of
salary data from the University of Arizona.<br>
It is built and maintained by one volunteer, [Chun Ly](https://astrochun.github.io).

This code is built purely with open-source software, specifically
[`python`](https://python.org), [`streamlit`](https://streamlit.io/),
[`bokeh`](https://bokeh.org/), and [`pandas`](https://pandas.pydata.org/).

The source code is available
[on GitHub!](https://github.com/astrochun/uarizona-salary-app)

If you have suggestions or encounter an issue, please feel free to submit an
issue request [on GitHub](https://github.com/astrochun/uarizona-salary-app/issues)!

As this is open source, we welcome contributions by
[forking](https://github.com/astrochun/uarizona-salary-app/fork) the repository, and
submitting a pull request!

This app was developed because I felt it was an important issue that
requires greater transparency.<br>
I maintain and develop this in my free time. With additional data, I hope to extend
this application's resources.

You can begin your data journey by selecting a "data view" on the sidebar:
    
 1. **Individual Search 🆕 : Find all salary data for individual(s)**
 2. Trends: General facts and numbers (e.g. number of employees,
    salary budget, etc.), for each fiscal year
 3. Salary Summary: Statistics and percentile salary data, includes salary histogram
 4. Highest Earners (Updated): Extract data above a minimum salary. Now you can select a given college/division
 5. College/Division Data: Similar to Salary Summary but extracted for each college(s)/division(s)
 6. Department Data: Similar to Salary Summary but extracted for each department(s)

Enjoy!<br>
&#8208; Chun 🌵

**Sponsorship:**<br>
If you find this website beneficial, please consider making a small
monetary donation either through GitHub or
[PayPal.Me](https://paypal.me/astrochun) (click on the Sponsor button).
This support ensures that this project will:

1. Grow to provide more data, tools, and visualizations, and
2. Continue to be an open-source transparent resource for our community.

_Your donations will be highlighted below._

This project has been sponsored by these gracious folks:

_Continued donors_:

1. [Casper da Costa-Luis](https://cdcl.ml/)

_One-time donors_:

1. John Moeller
2. Sedona Heidinger

**Sources:**<br>
The salary data are made available from the [The Daily WildCat](https://www.wildcat.arizona.edu/).
Links are available for:

 1. [FY2019-20](https://docs.google.com/spreadsheets/d/e/2PACX-1vTaAWak0pN6Jnulm95eTM7kIubvNNMPgYh3d6sCHN5W1tekpIktoMBoDKJeZhmAyI7ZzH1BAytEp_bV/pubhtml)
    (Google Sheets). Extracted and converted to a machine-readable version by Michael Clarkson.
 2. [FY2018-19](https://docs.google.com/spreadsheets/d/1d2wLowmL5grmsqTj-qFg2ke9k--s1gN_oEZ6kstSX6c/edit#gid=0) (Google Sheets)
 3. [FY2017-18](https://docs.google.com/spreadsheets/d/1jFmxbDx6FP5qk5KKbFBJ5RvS0R2_HEoCkaw83P_DUG0/edit#gid=2006091355) (Google Sheets)
 4. FY2011-12, FY2013-14, FY2014-15, FY2015-16, FY2016-17:
    These data were made available by Dr. David Le Bauer, obtained from
    older records from The Daily Wildcat (Google Sheets are currently not available/public)
    
Inflation data shown on Wage Growth plots were obtained from the U.S. Bureau of
Labor Statistics using their
[Consumer Price Index (CPI) Inflation Calculator](https://www.bls.gov/data/inflation_calculator.htm).
Calculations were either over a 1- or 2-year period
(depending on prior year's data) and adopted FY start
(e.g., July 2017 to July 2018).


CHANGELOG.md
# Changelog

## [v1.1.4](https://github.com/astrochun/uarizona-salary-app/tree/v1.1.4) (2021-10-24)

**Closed issues:**
 - Update streamlit, pandas, etc
   [#80](http://github.com/astrochun/uarizona-salary-app/issues/80)

**Merged pull requests:**
 - Update streamlit, pandas, etc
   [#81](http://github.com/astrochun/uarizona-salary-app/pull/81)


## [v1.1.3](https://github.com/astrochun/uarizona-salary-app/tree/v1.1.3) (2021-10-20)

**Implemented enhancements:**
 - Illustrate state funding trends over time
   [#79](http://github.com/astrochun/uarizona-salary-app/pull/79)

**Closed issues:**
 - Illustrate state funding trends over time
   [#78](http://github.com/astrochun/uarizona-salary-app/issues/78)


## [v1.1.2](https://github.com/astrochun/uarizona-salary-app/tree/v1.1.2) (2021-10-19)

**Implemented enhancements:**
 - Include comparison against CPI on Individual Search
   [#76](http://github.com/astrochun/uarizona-salary-app/pull/76)

**Closed issues:**
 - Include comparison against CPI on Individual Search
   [#74](http://github.com/astrochun/uarizona-salary-app/issues/74)


## [v1.1.1](https://github.com/astrochun/uarizona-salary-app/tree/v1.1.1) (2021-10-18)

**Implemented enhancements:**
 - Viz: Add inflation line for each year on wage growth plot
   [#73](http://github.com/astrochun/uarizona-salary-app/pull/73)

**Closed issues:**
 - Add .gitpod.yml for GitPod PyCharm support
   [#69](http://github.com/astrochun/uarizona-salary-app/issues/69)
 - Viz: Add inflation line for each year on wage growth plot
   [#72](http://github.com/astrochun/uarizona-salary-app/issues/72)

**Merged pull requests:**
 - Add .gitpod.yml for GitPod PyCharm support
   [#71](http://github.com/astrochun/uarizona-salary-app/pull/71)


## [v1.1.0](https://github.com/astrochun/uarizona-salary-app/tree/v1.1.0) (2021-10-06)

**Implemented enhancements:**
 - Add Wage Growth data view [#70](http://github.com/astrochun/uarizona-salary-app/pull/70)

**Closed issues:**
 - Add Wage Growth data view [#46](http://github.com/astrochun/uarizona-salary-app/issues/46)


## [v1.0.2](https://github.com/astrochun/uarizona-salary-app/tree/v1.0.2) (2021-10-03)

**Implemented enhancements:**
 - Feature: Add average Y2Y growth on individual search
   [#68](http://github.com/astrochun/uarizona-salary-app/pull/68)

**Closed issues:**
 - Update sponsor list for John Moeller [#65](http://github.com/astrochun/uarizona-salary-app/issues/65)
 - Feature: Add average Y2Y growth on individual search
   [#67](http://github.com/astrochun/uarizona-salary-app/issues/67)

**Merged pull requests:**
 - Update sponsor list for John Moeller [#66](http://github.com/astrochun/uarizona-salary-app/pull/66)


## [v1.0.1](https://github.com/astrochun/uarizona-salary-app/tree/v1.0.1) (2021-08-15)

**Implemented enhancements:**
 - Add ability in Individual Search view to search by department
   [#64](http://github.com/astrochun/uarizona-salary-app/pull/64)

**Closed issues:**
 - Add ability in Individual Search view to search by department
   [#63](http://github.com/astrochun/uarizona-salary-app/issues/63)


## [v1.0.0](https://github.com/astrochun/uarizona-salary-app/tree/v1.0.0) (2021-07-18)

**Implemented enhancements:**
 - Add individual search page; Improvements to Highest Earners data view
   [#62](http://github.com/astrochun/uarizona-salary-app/pull/62)

**Fixed bugs:**
 - Limit Highest Earners [#56](http://github.com/astrochun/uarizona-salary-app/issues/56)

**Closed issues:**
 - Add individual search page [#47](http://github.com/astrochun/uarizona-salary-app/issues/47)
 - Format Highest Earner page for currency, FTE and State Fund Ratio
   [#61](http://github.com/astrochun/uarizona-salary-app/issues/61)


## [v0.3.0](https://github.com/astrochun/uarizona-salary-app/tree/v0.3.0) (2021-07-16)

**Implemented enhancements:**
 - Use unique identifier for individuals for user-based views
   [#60](http://github.com/astrochun/uarizona-salary-app/pull/60)

**Closed issues:**
 - Add information about sponsors
   [#57](https://github.com/astrochun/uarizona-salary-app/issues/57)
 - Use unique identifier for individuals for user-based views
   [#59](https://github.com/astrochun/uarizona-salary-app/issues/59)

**Merged pull requests:**
 - Add information about sponsors [#58](https://github.com/astrochun/uarizona-salary-app/pull/58)


## [v0.2.1](https://github.com/astrochun/uarizona-salary-app/tree/HEAD) (2021-06-21)

**Implemented enhancements:**
 - Add sources to About page; Add copyright information; Simplify about page with TL;DR section
   [#51](http://github.com/astrochun/uarizona-salary-app/pull/51)
 - Add other salary data; Update About page for data views
   [#49](http://github.com/astrochun/uarizona-salary-app/pull/49)
 - Main module is too large, refactor for easier readability
   [#55](http://github.com/astrochun/uarizona-salary-app/pull/55)

**Closed issues:**
 - Add sources to About page [#39](http://github.com/astrochun/uarizona-salary-app/issues/39)
 - Add copyright information [#45](http://github.com/astrochun/uarizona-salary-app/issues/45)
 - Simplify about page with TL;DR section [#50](http://github.com/astrochun/uarizona-salary-app/issues/50)
 - Add other salary data [#48](http://github.com/astrochun/uarizona-salary-app/issues/48)
 - Update About page for data views [#52](http://github.com/astrochun/uarizona-salary-app/issues/52)
 - Main module is too large, refactor for easier readability
   [#53](http://github.com/astrochun/uarizona-salary-app/issues/53)


## [v0.2.0](https://github.com/astrochun/uarizona-salary-app/tree/v0.2.0) (2021-06-10)

**Implemented enhancements:**
 - Add Trends data view [#44](http://github.com/astrochun/uarizona-salary-app/pull/44)

**Closed issues:**
 - Add Trends data view [#43](http://github.com/astrochun/uarizona-salary-app/issues/43)


## [v0.1.11](https://github.com/astrochun/uarizona-salary-app/tree/v0.1.11) (2021-06-06)

**Implemented enhancements:**

salary_app/main.py
#!/usr/bin/env python3
import argparse

import streamlit as st
from streamlit.components.v1 import html

import pandas as pd

from constants import COLLEGE_NAME, FY_LIST, TITLE
import sidebar
import views


@st.cache
def load_data(local: str = ''):
    """Load data"""
    if local:
        print("Loading data from local source")
    else:
        print("Loading data from Google Drive")

    file_id = {
        'FY2019-20': '1d2l29_T-mOh05bglPlwAFlzeV1PIkRXd',
        'FY2018-19': '1paxrUyW1wZuK3bjSL_L7ckKEC6xslZJe',
        'FY2017-18': '1AnRaPpbRTLVyqdeqe6vkPMYgbNnw9zia',
        'FY2016-17': '1rXBuuXit5oWKtfnA05gsNtsWAyESeIs2',
        'FY2014-15': '1ZANVDr6Kw40MJYiOENWbLMTFEMWyf7f4',
        'FY2013-14': '1rQ8A2CIVhDYu0lESKVh72h6VUd8gIEFl',
        'FY2011-12': '1fQOzEHiOvc_H1NcLMlK3KVV1DJkRbRuX',
    }

    data_dict = {}
    for year in FY_LIST:
        year_split = year.split(' ')[0]
        if not local:
            url = f'https://drive.google.com/uc?id={file_id[year_split]}'
        else:
            url = f'{local}/{year_split}_clean.csv'
        data_dict[year_split] = pd.read_csv(url)

    # Get unique.csv
    if not local:
        unique_url = 'https://drive.google.com/uc?id=1-2aFLO1nbPWT02y8N8Suue0Gg2FisWub'
    else:
        unique_url = f'{local}/unique.csv'
    unique_df = pd.read_csv(unique_url)

    return data_dict, unique_df


@st.cache
def header_buttons() -> str:
    """Return white-background version of GitHub Sponsor button"""
    file_name = "assets/header_buttons.html"
    with open(file_name, 'r', encoding='utf-8') as f:
        buttons_html = f.read().replace('\n', '')

    return buttons_html


def main(bokeh=True, local: str = ''):
    st.set_page_config(page_title=f'{TITLE} - sapp4ua', layout='wide',
                       initial_sidebar_state='auto')

    # Display GitHub Sponsor at top
    buttons_html = header_buttons()
    html(buttons_html, width=600, height=40)

    st.title(TITLE)
    st.markdown(
        '''
        <style>
        [data-testid="stSidebar"][aria-expanded="true"] > div:first-child {
            width: 255px;
        }
        [data-testid="stSidebar"][aria-expanded="false"] > div:first-child {
           width: 250px;
           margin-left: -255px;
        }
        </style>
        <style>
        #MainMenu {visibility: hidden;}
        </style>
        
        <style>
        footer { visibility: hidden; }
        footer:after {
          content:'Copyright © 2021 Chun Ly. All rights reserved.';
          visibility: visible;
          display: block;
          position: relative;
          text-align: center;
          padding: 5px;
          top: 2px;
        }
        </style>        
        ''',
        unsafe_allow_html=True
    )

    # Load data
    data_dict, unique_df = load_data(local=local)

    # Sidebar, select data view
    view_select = sidebar.select_data_view()

    df = None

    # Sidebar FY selection
    fy_select = ''
    if view_select not in ['About', 'Trends', 'Individual Search']:
        fy_select = sidebar.select_fiscal_year(view_select)

        # Select dataframe
        df = data_dict[fy_select]
        st.sidebar.text(f"{fy_select} data imported!")

        if view_select == 'Wage Growth':
            expect_prev_year = 'FY' + "-".join(
                [str(int(v) - 1) for v in fy_select.replace('FY', '').split('-')]
            )
            if expect_prev_year in ['FY2012-13', 'FY2015-16']:
                st.sidebar.warning(
                    f"Data from previous year ({expect_prev_year}) not available. "
                    "This is a two-year comparison.")

    # Select pay rate conversion
    pay_norm = 1  # Default: Annual = 1.0
    if view_select not in ['About', 'Highest Earners', 'Individual Search']:
        pay_norm = sidebar.select_pay_conversion(
            fy_select, pay_norm, view_select
        )

    if view_select == 'About':
        views.about_page()

    if view_select == 'Trends':
        views.trends_page(data_dict, pay_norm)

    if view_select == 'Salary Summary':
        views.salary_summary_page(df, pay_norm, bokeh=bokeh)

    if view_select == 'Highest Earners':
        views.highest_earners_page(df)

    # Select by College Name
    if view_select == 'College/Division Data':
        views.subset_select_data_page(df, COLLEGE_NAME, 'college',
                                      pay_norm, bokeh=bokeh)

    # Select by Department Name
    if view_select == 'Department Data':
        views.subset_select_data_page(df, 'Department', 'department',
                                      pay_norm, bokeh=bokeh)

    if view_select == 'Individual Search':
        views.individual_search_page(data_dict, unique_df)

    if view_select == 'Wage Growth':
        views.wage_growth_page(data_dict, fy_select, pay_norm,
                               bokeh=bokeh)


if __name__ == '__main__':

    parser = argparse.ArgumentParser("Streamlit script")
    parser.add_argument('--local', default='', help='Local path to specify')
    args = parser.parse_args()

    main(bokeh=True, local=args.local)
